### ISSUE TITLE
Icons from KML not working

### ISSUE THREAD

**Issue by mediafreakch** [openlayers/openlayers#13113]
**Describe the bug**

On earth.google.com I added two placemarks on a map. For one, I used a standard Google Earth Icon. For the other, I uploaded a custom icon, which then ends up as a dataUri in the exported KML file.

Neither of them is shown properly when importing the KML into openlayers. Instead, the default yellow push-pins are shown.

**To Reproduce**

1. Open https://codesandbox.io/s/kml-forked-nfovz
2. Look at the map

**Expected behavior**

![Bildschirmfoto 2021-12-14 um 08 00 30](https://user-images.githubusercontent.com/777093/145948807-a01060f9-66b1-4c4b-b20a-7a44545f50af.png)

**Actual behavior**

![Bildschirmfoto 2021-12-14 um 08 00 38](https://user-images.githubusercontent.com/777093/145948819-ffc05e08-a0be-4373-b97e-4ce00c0a4948.png)

**Comment by jipexu** [comment]
Hi 
https://codesandbox.io/s/kml-forked-nfovz
this seems work well  as the yellow pushpin are the default...
i dont see any code to stylise  them ???

**Comment by mediafreakch** [comment]
Ok then maybe I am unaware of how this is supposed to work... Shouldn't the icons referenced in the KML be transformed to openlayer icon styles automatically?

**Comment by jipexu** [comment]
i think perhaps this can help
https://github.com/openlayers/openlayers/issues/12437

**Comment by mediafreakch** [comment]
No, there's no indication of a CORS error anywhere. You can look at the [codesandbox[(https://codesandbox.io/s/kml-forked-uumk2?file=/test.kml) and see for yourself.  
You can also inspect the KML file in the sandbox, it references two icons:

1. https://earth.google.com/earth/rpc/cc/icon?color=1976d2&id=2061&scale=4
2. ```data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABGdBTUEAALGPC_xhBQAAACBjSFJNAAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAABmJLR0QAAAAAAAD5Q7t_AAAACXBIWXMAAABgAAAAYADwa0LPAAAAhklEQVRIx-2RwQ2DMBAE16nDPKB0P1AqgAIgtaCkieHjSImF4QQ8b14n3Wl3dSs5VwBeHDPtaYQDAyxBQghVnceG6OcbrVgNkqKkJs9_QTJvy1uelVfEn5umctNbDDpgOWGwAJ3lpQJaIBUCAxCz-FjsEtB6yV6yl-wl31VywWzQnww3Tp0VnZpacqu5CHUAAAAldEVYdGRhdGU6Y3JlYXRlADIwMjEtMTItMTNUMTU6NDA6MzIrMDA6MDCN4HcZAAAAJXRFWHRkYXRlOm1vZGlmeQAyMDIxLTEyLTEzVDE1OjQwOjMyKzAwOjAw_L3PpQAAAABJRU5ErkJggg```

Both are not displayed.  
The issue you linked to did not help (setting `crossorigin: null`). I don't think it's related.

**Comment by mike-000** [comment]
The output is similar when your test.kml is dropped onto Google Earth.  If you are expecting something else that suggests an error or non-standard setting in the file.

![image](https://user-images.githubusercontent.com/49240900/145979469-5a85ea41-7169-41bd-b50b-73089425c2a4.png)

**Comment by jipexu** [comment]
@mediafreakch 
ah now  i see the icon are not from url but directly by data:image .... huummm

**Comment by mediafreakch** [comment]
Interesting, it's true that with Google Earth Pro (7.3.4) on OSX, the icons appear with the default style only too. However on the web version (https://earth.google.com/web/) they appear correctly. Maybe it uses a slightly newer KML syntax?

And yes, one icon references a URL from Google, the other is inline using a dataUri

![Bildschirmfoto 2021-12-14 um 11 59 55](https://user-images.githubusercontent.com/777093/145985879-c4e24d06-823e-4fb7-9966-38690573926a.png)

**Comment by mediafreakch** [comment]
So seems that the KML parser does not support `<gx:CascadingStyle />` tags. See same issue here https://gis.stackexchange.com/a/367696 .

I am also unsure, why Google Earth Web exports the KML wrapping the Style tags in such a tag. According to the specs, this doesn't exist: https://developers.google.com/kml/documentation/kmlreference#kml-extension-namespace-and-the-gx-prefix

I created another example [here](https://codesandbox.io/s/kml-debugging-ssbzb?file=/test.kml), where I manually unwrapped the style tags. Now they are children of `<Document />`. At least the icon referenced by URL now is displayed correctly. I'll try to come up with a workaround to do this automatically.

However the icon included as dataUri still doesn't work. Any ideas on where the problem is for this one?

**Comment by mediafreakch** [comment]
So it seems the base64 encoded URL was wrong. There were two `==` missing at the end. Seems that the export from Google Earth Web somehow strips them. After adding them, everything works as expected. I updated the example: https://codesandbox.io/s/kml-debugging-ssbzb?file=/test.kml

I'll try to find a way to file a bug for Google Earth Web.

**Comment by jipexu** [comment]
ah cool yes it's works on your example with the dataUri ... good to know because the goggle doc seems very poor on this
`An HTTP address or a local file specification used to load an icon.`
or may be   `local file specification` include datUri

**Comment by mediafreakch** [comment]
I played around a bit and came up with a quick and dirty example on how to handle Google Earth Web KMLs automatically.
The approach subclasses `KML()`. Use `GoogleEarthWebKML` instead of `KML` in the format options of the Source.

```
class GoogleEarthWebKML extends KML {
  getType() {
    return 'text'
  }

  readFeature(source, options) {
    return super.readFeature(this.handleBadTags(source), options)
  }

  readFeatures(source, options) {
    return super.readFeatures(this.handleBadTags(source), options)
  }

  handleBadTags(rawString) {
    let result = rawString

    // remove "cascadingStyle" Tags
    result = rawString.replace(/<.*?cascadingstyle.*?kml:id="(.+)">\s*<style>/gmi, (a, b) => {
        return "<Style id=\"" + b + "\">";
    });
    result = result.replace(/<\/Style>\s*<\/.*?cascadingstyle>/gmi, "</Style>");

    // fix dataURIs (replace -_ with +/ and add padding if necessary)
    result = result.replace(base64Regex, (a, b) => {
      const withPadding = b.length % 4 !== 0 ? a += Array.from({ length: 4 - b.length % 4 }, () => '=').join('') : a
      return withPadding.replaceAll('-', '+').replaceAll('_', '/')
    })

    return result;
  }
}
```

You can see a working example here: https://codesandbox.io/s/kml-earth-fix-u7drq?file=/main.js
